### DISCUSSION TITLE
RAM management

### DISCUSSION THREAD

**Discussion by odudex** [selfcustody/krux#229] [Ideas]
After an user reported having RAM issues I've been refactoring parts of the code, splitting files and loading modules "on demand", inside of functions, instead of at the header of files. Another side effect of "on demand" module loading is Krux boots faster. This approach can have some miliseconds lags during usage, but nothing annoying or, in worst case, that can't be mitigated by a "loading" message. I also would like to unload modules, specially those use only once, but tests and literature showed this cannot be done efficiently in micropython. I believe at the current state we won't have RAM issues, but I started this discussion to check if there's any other implication on the "on demand module loading" approach, collect ideas, and make contributors aware RAM should be watched closely for new features.

**Comment by tadeubas** [comment]
I think that you said: 'that CAN be mitigated by a "loading" message'. Plz share your experience on tests and the literature that shows 'this cannot be done efficiently' (the memory management, deleting variables and removing loaded modules)

**Comment by odudex** [comment]
The "literature" maybe was to too fancy term for micropython discussions 😁.
Fact is when deleting variables and loaded modules, the available RAM won't get back to the state it was before, even if you hunt down and delete all chain of imported modules, it seems a lot of "junk" remains.
Different approaches can be made, but none will do a perfect clean up.
https://forum.micropython.org/viewtopic.php?t=5302

I'm sure there's still a lot that can be optimized in this regard on Krux, but we will have to study and do more tests.

**Comment by ghost** [comment]
@odudex Was the RAM issue related to PSBT signing by chance? If so, then I think this issue should be reopened: https://github.com/selfcustody/krux/issues/2

I noticed early on when the firmware was larger that if I signed a transaction with many inputs, I'd run out of memory. I went looking for a solution and found that embit had one: PSBTView. I believe it pages large transactions into and out of memory rather than loading the entire thing at once. We should probably switch to using it regardless...

Also, something to keep in mind: the entire firmware is loaded into memory, so it's a good idea to keep it as small as possible as well.

**Comment by tadeubas** [comment]
Yes it is exactly this problem! I saw you reopened #2 👏 , this big transaction problem needs to be addressed...

**Comment by odudex** [comment]
On "medium size" transactions I was using to test the issue was solved for now. For further investigations, I think UR may use a lot of RAM too. Also, we could still increase [heap size](https://github.com/selfcustody/MaixPy/blob/85dba62b83054879772aa27524bc2cf53e37bf4b/projects/maixpy_amigo_tft/config_defaults.mk#L56C1-L56C35), couldn't we?

**Reply by ghost** [reply]
Yes, I need to do some profiling on the [urtypes library](https://github.com/selfcustody/urtypes) and the [UR library](https://github.com/selfcustody/foundation-ur-py/tree/8308a4a6399cc361a3caaef92f0924ddc0d7e102) from Foundation Devices to see how we can improve RAM usage. Will file an issue for this
